### PR TITLE
Slack -- add ts to incoming messages so that bot can send reactions

### PR DESCRIPTION
**Pull request by r3m0t**
On Slack a useful alternative to an "OK" reply is a reaction.

### CODE REVIEW

File: errbot/backends/slack.py [hunk 1]
```diff
@@ -838,6 +838,26 @@ def build_reply(self, mess, text=None, private=False):
...
+                                  timestamp=mess.extras['ts'],
+                                  name=reaction)
+        if result['ok']:
+            return
+        elif result.get('error') == 'invalid_name':
+            raise ValueError(result['error'], 'No such emoji', reaction)
+        elif result.get('error') in ('no_reaction', 'already_reacted'):
+            raise LookupError(result['error'])
+        else:
+            raise Exception(result['error'])
```
Comment (zoni) [inline errbot/backends/slack.py]: Can you raise [SlackAPIResponseError](https://github.com/r3m0t/errbot/blob/64b89797abad251fcd3eee4f7f9526fe71d18e62/errbot/backends/slack.py#L89-L98) here instead?

File: errbot/backends/slack.py [hunk 2]
```diff
@@ -838,6 +838,26 @@ def build_reply(self, mess, text=None, private=False):
...
+        result = self.sc.api_call(api_call,
+                                  channel=mess.to.channelid,
+                                  timestamp=mess.extras['ts'],
+                                  name=reaction)
+        if result['ok']:
+            return
+        elif result.get('error') == 'invalid_name':
+            raise ValueError(result['error'], 'No such emoji', reaction)
+        elif result.get('error') in ('no_reaction', 'already_reacted'):
+            raise LookupError(result['error'])
```
Comment (zoni) [inline errbot/backends/slack.py]: I think a `ValueError` makes more sense here as well.

File: errbot/backends/slack.py [hunk 3]
```diff
@@ -838,6 +838,26 @@ def build_reply(self, mess, text=None, private=False):
...
+        return self._react('reactions.add', mess, reaction)
+
+    def remove_reaction(self, mess, reaction):
+        return self._react('reactions.remove', mess, reaction)
+
+    def _react(self, api_call, mess, reaction):
+        result = self.sc.api_call(api_call,
+                                  channel=mess.to.channelid,
+                                  timestamp=mess.extras['ts'],
+                                  name=reaction)
```
Comment (zoni) [inline errbot/backends/slack.py]: I would rather make use of the [self.api_call()](https://github.com/r3m0t/errbot/blob/64b89797abad251fcd3eee4f7f9526fe71d18e62/errbot/backends/slack.py#L303-L334) helper function here.

**Comment by r3m0t** [comment]
I will edit the docs, basically we have a command `!notify [task-name]` that notifies the user privately when the task (that already exists) completes, a few minutes later. It looks odd without an immediate reply, but the replies clog up the channel. So, instead of the `notify` botcmd returning a message, it returns `None` and calls `self._bot.add_reaction(msg, 'ok_hand')`. There are other potential uses as well.

File: errbot/backends/slack.py (hunk 1 above)
Comment (r3m0t) [inline errbot/backends/slack.py]: Sure, but I don't understand how plugins can import SlackAPIResponseError in case they want to catch it, given yapsy?

**Comment by r3m0t** [comment]
I couldn't build the docs in my local environment, so there might be an issue there. I also saved `ts` on outgoing messages, but it's almost impossible to use as even if you pass in a `Message` object, it gets replaced multiple times, and doesn't get returned. I'm sure other backends support editing messages as well, so perhaps it would be useful to return `Message` instead of `None` in some places? But, I wouldn't have any use for that feature as of now.

Here's my final test plugin:

```python
from errbot import BotPlugin, botcmd, Message
from time import sleep

class Example(BotPlugin):
    @botcmd
    def longcompute(self, mess, args):
        if self._bot.mode == "slack":
            self._bot.add_reaction(mess, "hourglass")
        else:
            yield "Finding the answer..."
        M = Message(body='hi', to=mess.frm)
        self._bot.send_message(M)
        self._bot.add_reaction(M, "ok_hand")
        sleep(10)
        yield "The answer is: 42"
        if self._bot.mode == "slack":
            self._bot.remove_reaction(mess, "hourglass")
```

**Comment by r3m0t** [comment]
@zoni I added another member to `Message.extras` on received messages which is a permalink to the message. Please re-review?

**Comment by zoni** [comment]
I will try to review this somewhere over the coming few days, currently really swamped which isn't leaving me much time. Please poke me if I end up taking too long to get back to this.

File: errbot/backends/slack.py [hunk 4]
```diff
@@ -505,7 +505,8 @@ def _message_event_handler(self, event):
                 msg.frm = SlackRoomBot(
                     self.sc,
                     bot_id=event['bot_id'],
-                    bot_username=event['username'],
+                    bot_username=event.get('username', 'bot'),  # SlackClient doesn't keep the bot data from the login
```
Comment (r3m0t) [inline errbot/backends/slack.py]: I will move this unrelated change

File: errbot/backends/slack.py [hunk 5]
```diff
@@ -1014,6 +1021,8 @@ def id(self):
             self._id = self._channel.id
         return self._id
 
+    channelid = id
```
Comment (r3m0t) [inline errbot/backends/slack.py]: this might be removable

**Comment by r3m0t** [comment]
@zoni happy new year! I probably have time next week to work on this. What would you like done if anything?

**Comment by r3m0t** [comment]
@zoni poke

File: errbot/backends/slack.py (hunk 4 above)
Comment (zoni) [inline errbot/backends/slack.py]: I think this can be removed, yes?

File: errbot/backends/slack.py (hunk 5 above)
Comment (zoni) [inline errbot/backends/slack.py]: It doesn't look like this is used anywhere, can you see if it's possible to remove it?

**Comment by rcoup** [comment]
Did this get dropped? Or merged in somewhere else?

**Comment by r3m0t** [comment]
I moved my fork into my employer's account. I'll open a new pull request.

**Pull request closed without merging**